pkg/datasource.go

```diff
@@ -503,105 +507,133 @@ func (o *OCIDatasource) processLogRecords(ctx context.Context, searchLogsReq Gra
+									// For all other keys treat them generically as string type
+								} else if key != LogSearchResultsField_Subject {
+									var stringFieldValue string
+									fieldDefn = nil
+
+									if stringFieldValue, ok = value.(string); ok {
+										// If the field value is non-zero length string then proceed to get/create the data
+										// field definition. But if the field value is a zero length string then skip
+										// creating the data field definition, this is to avoid creating a data field for a
+										// log record field that is always empty.
+										if len(stringFieldValue) > 0 {
+											fieldDefn = o.getCreateDataFieldElemsForField(mFieldDefns, numMaxResults,
+												key, key, ValueType_String)
+										}
+									} else {
+										o.logger.Debug("Error parsing string field value", "panelId", queryPanelId,
+											"refId", queryRefId, "key", key, "value", value)
+										fieldDefn = o.getCreateDataFieldElemsForField(mFieldDefns, numMaxResults,
+											key, key, ValueType_String)
+										stringFieldValue = "UNKNOWN"
+									}
+									if fieldDefn != nil {
+										fieldDefn.Values.([]*string)[indexCountPag] = &stringFieldValue
+									}
+								} // endif key name
+							} // for each field key in the logContent field
+
+						} else {
+							o.logger.Debug("Unable to get logContent map", "panelId", queryPanelId,
+								"refId", queryRefId, "row", rowCount)
+						}
 					} else {
-						o.logger.Debug("Unable to get logContent map", "panelId", queryPanelId,
-							"refId", queryRefId, "row", rowCount)
+						o.logger.Debug("Encountered log record without a logContent element",
+							"panelId", queryPanelId, "refId", queryRefId, "row", rowCount)
 					}
 				} else {
-					o.logger.Debug("Encountered log record without a logContent element",
-						"panelId", queryPanelId, "refId", queryRefId, "row", rowCount)
+					o.logger.Debug("Encountered row without a log record", "panelId", queryPanelId,
+						"refId", queryRefId, "row", rowCount)
 				}
-			} else {
-				o.logger.Debug("Encountered row without a log record", "panelId", queryPanelId,
-					"refId", queryRefId, "row", rowCount)
+				indexCountPag++
 			}
+
+		} else {
+			o.logger.Warn("Logging search query returned no results", "panelId", queryPanelId,
+				"refId", queryRefId)
+		}
+		if res.OpcNextPage != nil && numpage < MaxPagesToFetch {
+			// if there are more items in next page, fetch items from next page
+			request.Page = res.OpcNextPage
+			numpage++
+		} else {
+			o.logger.Debug("numpage :", "numpage", numpage)
+			o.logger.Debug("resultCountPag :", "resultCountPag", resultCountPag)
+			o.logger.Debug("Reducing data field values", "resultsCount", resultCountPag-1)
+			for _, dataFieldDefn := range mFieldDefns {
+				if dataFieldDefn.Type == ValueType_Time {
+					timeValuesSlice, _ := dataFieldDefn.Values.([]*time.Time)
+					dataFieldDefn.Values = timeValuesSlice[:resultCountPag-1]
```

Comment:
Unfortunately when I prototyped this logic I didn't understand that the upper range index is exclusive, i.e. the value at that index value is not included in the resulting slice. So the reduction of the upper range index value by one in this statement and the ones below needs to be removed. If not then one less row than was retrieved will be presented in Grafana, see this screen shot of a logs data panel where the maximum number of rows was returned:
![LogRecordsDataPanel](https://user-images.githubusercontent.com/37004987/177006235-9dfe8154-f7ee-4231-b52a-f994f67d8ebf.png)

Reply:
ok, I removed the "-1" and now it gives 10000 rows